pyomo/contrib/viewer/model_browser.py

```diff
@@ -0,0 +1,479 @@
+    """
+    def __init__(self, parent, o, ui_setup):
+        self.ui_setup = ui_setup
+        self.data = o
+        self.parent = parent
+        self.children = [] # child items
+        self.ids = {}
+        self.clear_cache() # cache is for calculated items
+        self.get_callback = {
+            "value": self._get_value_callback,
+            "lb": self._get_lb_callback,
+            "ub": self._get_ub_callback,
+            "expr": self._get_expr_callback}
+        self.set_callback = {
+            "value":self._set_value_callback,
+            "lb":self._set_lb_callback,
+            "ub":self._set_ub_callback,
+            "active":self._set_active_callback,
+            "fixed":self._set_fixed_callback}
+
+    def data_items():
+        """Iterate through chedren data items and this one"""
+        for i in self.children:
+            i.data_items()
+        yield self
+
+    def add_child(self, o):
+        """Add a child data item"""
+        item = ComponentDataItem(self, o, ui_setup=self.ui_setup)
+        self.children.append(item)
+        self.ids[id(o)] = item
+        return item
+
+    def clear_cache(self):
+        """Clear chache for calcuated items"""
+        self._cache_value = None
+        self._cache_lb = None
+        self._cache_ub = None
+
+    def calculate_children(self):
+        self.calculate()
+        for i in self.children:
+            i.calculate_children()
+
+    def calculate(self):
+        """Calculate items, applies to expressions and constraints"""
+        if isinstance(self.data, _ExpressionData):
+            self._cache_value = value(self.data)
+        if isinstance(self.data, _ConstraintData) and self.data.active:
+            try:
+                self._cache_value = value(self.data.body)
+                self._cache_lb = value(self.data.lower)
+                self._cache_ub = value(self.data.upper)
+            except:
+                pass
+
+    def get(self, a):
+        """Get an attribute"""
+        if a in self.get_callback:
+            return self.get_callback[a]()
+        else:
+            try:
+                return getattr(self.data, a)
+            except:
+                return None
+
+    def set(self, a, val):
+        """set an attribute"""
+        if a in self.set_callback:
+            return self.set_callback[a](val)
+        else:
+            try:
+                return setattr(self.data, a, val)
+            except:
+                return None
+
+    def _get_expr_callback(self):
+        if hasattr(self.data, "expr"):
+            return str(self.data.expr)
+        else:
+            return None
+
+    def _get_value_callback(self):
+        if isinstance(self.data, _VarData):   # prevent Pyomo warinig about
+            if self.data.value is None:       # value of non-numeric value
+                return None
+        if isinstance(self.data, (_VarData, _ParamData, float, int)):
+            try:
+                return value(self.data)
```

Comment:
There is an option called `exception` that can be passed to the value function to avoid an exception being thrown if components are not initialized and instead return `None`. I think you could use this option to drastically simplify the logic here. `value(self.data, exception=False)`

Reply:
Thanks that fixed it.  Gotta start reading documentation.